mesonbuild/modules/gnome.py

```diff
@@ -775,6 +775,10 @@ def gtkdoc(self, state, args, kwargs):
                 '--mode=' + mode]
         if namespace:
             args.append('--namespace=' + namespace)
+        gtkdoc_exe_wrapper = state.environment.cross_info.config["properties"].get('gtkdoc_exe_wrapper', None)
```

Comment:
Why are you using a custom property rather than the default exe wrapper here?

Reply:
We don't have a 'universal' wrapper script/binary to use. The wrapper script that gets used here is created specifically for the purpose of running transient gtk-doc binaries under emulation, and sets up library paths to look in and glib tweaks in a way that is specific to gtk-doc, and so is known and tested to work for that specific case.

Here's where the wrapper script is created:
http://git.yoctoproject.org/cgit/cgit.cgi/poky/tree/meta/classes/gtk-doc.bbclass#n40

Reply:
It would be okay for us to first check for the custom gtkdoc wrapper, and if that is not available, then use the 'standard' exe wrapper. But we would stick with the custom one.

Reply:
Is this something that should be sent to gtkdoc upstream instead? It seems like a general problem.

Reply:
I don't think so. Gtk-doc takes a command-line argument for the wrapper, and leaves the decision for what the wrapper should be doing to the build system. And the things that our wrapper is doing are very much specific to how Yocto does things - they're certainly not fixing any flaws in gtk-doc. I'm not sure what we could offer them.

Reply:
Thinking about this further we usually try to deal with these cases by having Meson call a custom wrapper script that has all the magic stuff in it so we don't have to build special cases for every system + dog in the world.

In this case you would have an entry in the cross file binaries section for gtk-doc, it would point to a shell script that does roughly this:

    #!/bin/sh
    /the/real/gtkdoc-binary --magic-args "$@"

That script could then be used with every other build system as well without needing to change them in any way. Would something like this be feasible?

Reply:
Yes, that would work. That's how we deal with gobject-introspection for instance: rather than call g-i tools directly, we substitute special wrappers that set up emulation magic. Meson (or autotools or cmake) doesn't even notice anything, as it determines the names of the g-i tools to run via pkg-config (see my other patch). Unfortunately I don't think gtk-doc has a similar pkg-config facility for tool names (I'll check).

Reply:
You can override any executable name you want in the cross file. If you can't override those tools then it is a bug (similarly we should permit overriding Qt's tools which we currently don't).

Reply:
Thanks, I'll see what can be done with these overrides. Certainly better to set things in config than patch source code.

Reply:
I have looked, and sadly, the substitution from cross file is not going to work. We need to substitute gobject-introspection tools and gtk-doc tools that meson is calling internally, and their names are hardcoded inside meson's source code. Cross file substitutions only work for find_program statements in meson.build files.

Reply:
Yes, and that is a bug we need to fix. All these should be overridable.

Reply:
I suppose the first step would be to make find_program() in meson source code use the same implementation as find_program() used for meson.build files. That the two are different had me confused for a while.

Otherwise, if I separate the wrapper substitution stuff, and re-submit the rest, would you be interested to take another look at it? We really do want to minimize the amount of patches to carry, as they can be a pain to rebase.